source/screenshotcreator.ts

```diff
@@ -0,0 +1,167 @@
+ *
+ * ```
+ * const screenshotCreator = new ScreenshotCreator(context);
+ * const image = screenshotCreator.capture(accumulateFramebuffer, gl.COLOR_ATTACHMENT0);
+ * ScreenshotCreator.download(image);
+ * ```
+ */
+export class ScreenshotCreator {
+    /**
+     * Read-only access to the objects context, used to get WebGL API access.
+     */
+    protected _context: Context;
+
+    /**
+     * Downloads the given image as png.
+     * @param imageData - Image to download.
+     */
+    static download(imageData: ImageData): void {
+        const canvas = document.createElement('canvas');
+        const context = canvas.getContext('2d')!;
+
+        canvas.width = imageData.width;
+        canvas.height = imageData.height;
+
+        context.putImageData(imageData, 0, 0);
+
+        const image = new Image();
+        image.src = canvas.toDataURL();
+
+        const link = document.createElement('a');
+        link.download = 'screenshot.png';
+        link.href = image.src;
+        link.click();
+    }
+
+    constructor(context: Context) {
+        this._context = context;
+    }
+
+    /**
+     * Captures the gl.BACK buffer of the default framebuffer and writes it into the given framebuffer. Sets the color
+     * buffer source for the default framebuffer to gl.BACK.
+     * @param rect - Part of the framebuffer to capture.
+     * @param buffer - Buffer to write the data into.
+     */
+    protected captureDefaultFramebufferIntoBuffer(rect: ScreenshotCreator.Rect, buffer: Uint8Array): void {
+        const gl = this._context.gl;
+
+        gl.readBuffer(gl.BACK);
+        gl.readPixels(rect.x, rect.y, rect.width, rect.height, gl.RGBA, gl.UNSIGNED_BYTE, buffer);
+    }
+
+    /**
+     * Captures the given attachment of the given framebuffer. Sets the color buffer source to the given attachment.
+     * @param framebuffer - Framebuffer to capture.
+     * @param attachment - Attachment to capture.
+     * @param rect - Part of the framebuffer to capture.
+     * @param buffer - Buffer to write the data into.
+     */
+    protected captureFramebufferIntoBuffer(framebuffer: Framebuffer, attachment: GLenum,
+        rect: ScreenshotCreator.Rect, buffer: Uint8Array): void {
+        const gl = this._context.gl;
+
+        const texture = framebuffer.texture(attachment);
+        assert(texture !== undefined, `Framebuffer does not have given attachment ${attachment}.`);
+        assert(texture!.format === gl.RGBA, `Unsupported texture format ${texture!.format}.`);
+        assert(texture!.type === gl.UNSIGNED_BYTE || texture!.type === gl.FLOAT || texture!.type === gl.HALF_FLOAT,
+            `Unsupported texture type ${texture!.type}.`);
+
+        gl.readBuffer(attachment);
+
+        if (texture!.type === gl.FLOAT || texture!.type === gl.HALF_FLOAT) {
+            const tempBuffer = new Float32Array(rect.width * rect.height * 4);
+            gl.readPixels(rect.x, rect.y, rect.width, rect.height, gl.RGBA, gl.FLOAT, tempBuffer);
```

Comment:
I'm not sure if this is standard conform, but it works in both Chrome and Firefox. Since Typescript has no 16-bit float buffer this is the easiest solution.